gflownet/envs/cube.py

```diff
@@ -0,0 +1,1317 @@
+
+        Returns
+        -------
+        self.state : list
+            The sequence after executing the action
+
+        action : int
+            Action executed
+
+        valid : bool
+            False, if the action is not allowed for the current state, e.g. stop at the
+            root state
+        """
+        if self.done:
+            return self.state, action, False
+        if action == self.eos:
+            assert self.state != self.source
+            self.done = True
+            self.n_actions += 1
+            return self.state, self.eos, True
+        # Otherwise perform action
+        else:
+            self.n_actions += 1
+            self._step(action, backward=False)
+            return self.state, action, True
+
+    # TODO: make generic for continuous environments
+    def step_backwards(
+        self, action: Tuple[int, float]
+    ) -> Tuple[List[float], Tuple[int, float], bool]:
+        """
+        Executes backward step given an action. An action is the absolute decrement of
+        each dimension.
+
+        Args
+        ----
+        action : tuple
+            Action to be executed. An action is a tuple of length n_dim, with the
+            absolute decrement for each dimension.
+
+        Returns
+        -------
+        self.state : list
+            The sequence after executing the action
+
+        action : int
+            Action executed
+
+        valid : bool
+            False, if the action is not allowed for the current state, e.g. stop at the
+            root state
+        """
+        # If done is True, set done to False, increment n_actions and return same state
+        if self.done:
+            assert action == self.eos
+            self.done = False
+            self.n_actions += 1
+            return self.state, action, True
+        # Otherwise perform action
+        else:
+            assert action != self.eos
+            self.n_actions += 1
+            self._step(action, backward=True)
+            return self.state, action, True
+
+    def get_grid_terminating_states(self, n_states: int) -> List[List]:
+        n_per_dim = int(np.ceil(n_states ** (1 / self.n_dim)))
+        linspaces = [np.linspace(0, self.max_val, n_per_dim) for _ in range(self.n_dim)]
+        states = list(itertools.product(*linspaces))
+        # TODO: check if necessary
+        states = [list(el) for el in states]
+        return states
+
+    def get_uniform_terminating_states(
+        self, n_states: int, seed: int = None
+    ) -> List[List]:
+        rng = np.random.default_rng(seed)
+        states = rng.uniform(low=0.0, high=self.max_val, size=(n_states, self.n_dim))
+        return states.tolist()
+
+    #     # TODO: make generic for all environments
+    def sample_from_reward(
+        self, n_samples: int, epsilon=1e-4
+    ) -> TensorType["n_samples", "state_dim"]:
+        """
+        Rejection sampling  with proposal the uniform distribution in
```

Comment:
Uhm, I don't think so. Being a bit more verbose it would say: "Rejection sampling where the proposal distribution is the uniform distribution in..."

This is being changed in a separate (WIP) [branch](https://github.com/alexhernandezgarcia/gflownet/tree/sample-from-reward). I will make the docstrings clearer there.